From 660c2dca9f7904e61d5e9341e22ab8c5ec8794c3 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 20 Jan 2014 21:04:00 -0500
Subject: [PATCH] Cap Entity Collisions

Limit a single entity to colliding a max of 10 times per tick.
This will alleviate issues where animals are hoarded in 1x1 pens.
---
 src/main/java/net/minecraft/server/EntityLiving.java | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index b4dab47..a51521a 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -1487,15 +1487,17 @@ public abstract class EntityLiving extends Entity {
     protected void bo() {
         List list = this.world.getEntities(this, this.boundingBox.grow(0.20000000298023224D, 0.0D, 0.20000000298023224D));
 
+        int count = 0;boolean skip = !(this instanceof EntityPlayer) && this.ticksLived % 2 == 0; // EMC
         if (this.R() && list != null && !list.isEmpty()) { // Spigot: Add this.R() condition
             for (int i = 0; i < list.size(); ++i) {
                 Entity entity = (Entity) list.get(i);
 
                 // TODO better check now?
                 // CraftBukkit start - Only handle mob (non-player) collisions every other tick
-                if (entity instanceof EntityLiving && !(this instanceof EntityPlayer) && this.ticksLived % 2 == 0) {
+                if (entity instanceof EntityLiving && skip) { // EMC
                     continue;
                 }
+                if (count++ > 10) { return; } // EMC
                 // CraftBukkit end
 
                 if (entity.S()) {
-- 
1.8.3.2

